### PR TITLE
Circuit API

### PR DESCRIPTION
**Pull request by imeckler**

### CODE REVIEW

File: kimchi/src/prover.rs
Comment (mimoo) [inline kimchi/src/prover.rs:99]: we should just replace all of these with `if cfg!(debug_assertions)`.

File: kimchi/src/prover.rs
Comment (mimoo) [inline kimchi/src/prover.rs:180]: `c.zip(b).unwrap().map(|(g, b)|`

that's a lot of one-letter variable names :p

File: circuit-construction/src/lib.rs
Comment (mimoo) [inline circuit-construction/src/lib.rs:122]: we already have this type as `CircuitGate`

File: circuit-construction/src/lib.rs
Comment (mimoo) [inline circuit-construction/src/lib.rs:761]: I think these gadgets should live within their respective files, not here. Poseidon, for example, already has a gadget here https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/polynomials/poseidon.rs#L95

File: circuit-construction/src/lib.rs
Comment (querolita) [inline circuit-construction/src/lib.rs:809]: where is the wiring handled?

**Review by querolita — COMMENTED** [review]
Awesome line of work 🦾

File: circuit-construction/src/lib.rs
Comment (mimoo) [reply circuit-construction/src/lib.rs:143]: +1, that's not a great name

File: circuit-construction/benches/example_proof.rs
Comment (o1-david) [inline circuit-construction/benches/example_proof.rs:69]: I don't think this should be part of the circuit :o and we should continue to do it transparently

File: circuit-construction/src/lib.rs
Comment (jspada) [reply circuit-construction/src/lib.rs:761]: Agree-- it's nice to have clean separation and all of a gate's code in a single file or module.  I think the proof-system's hooks into the gate should be as clean as possible and not have to know internal details about the gate.

File: circuit-construction/src/lib.rs
Comment (jspada) [reply circuit-construction/src/lib.rs:143]: Good one.

File: kimchi/src/prover.rs
Comment (o1-david) [inline kimchi/src/prover.rs:83]: why are you doing this?

Comment (imeckler) [reply kimchi/src/prover.rs:83]: If you want we could make it a separate method with a scary name? It would be nice to merge it in though

Comment (imeckler) [reply kimchi/src/prover.rs:83]: If you want we could make it a separate method with a scary name? My preference would be to merge it but it's OK if we don't

File: groupmap/src/lib.rs
Comment (mimoo) [reply groupmap/src/lib.rs:31]: agree, but for different PR/issue :o

File: kimchi/src/prover.rs
Comment (mimoo) [inline kimchi/src/prover.rs:190]: I guess I don't understand why removing these and adding the zk rows manually on the circuit side, is this a question of control as well for the bba thing?

File: circuit-construction/src/lib.rs
Comment (mimoo) [reply circuit-construction/src/lib.rs:459]: resolving these as I'm trying to get this in, we can propose refactors in other PRs :o

File: kimchi/src/prover.rs
Comment (mimoo) [reply kimchi/src/prover.rs]: I added a wrapping `Option<[Option<.>; COLUMNS]>` for now

**Comment by mimoo** [comment]
cc @imeckler can you take a look at this now?

**Review by mimoo — APPROVED** [review]
This looks good to me now, @imeckler can you take a look and cross approve before we merge it?

**Comment by imeckler** [comment]
I cannot approve as I'm the original author, but I will merge :+1:

**Pull request merged**
